src/Microsoft.HttpRepl.Telemetry/Telemetry.cs

```diff
@@ -0,0 +1,202 @@
+
+        public static bool SkipFirstTimeExperience => GetEnvironmentVariableAsBool(HttpRepl.Telemetry.FirstTimeUseNoticeSentinel.SkipFirstTimeExperienceEnvironmentVariableName, false);
+
+        public static bool IsRunningInDockerContainer => GetEnvironmentVariableAsBool("DOTNET_RUNNING_IN_CONTAINER", false);
+
+        private static bool GetEnvironmentVariableAsBool(string name, bool defaultValue = false)
+        {
+            var str = Environment.GetEnvironmentVariable(name);
+            if (string.IsNullOrEmpty(str))
+            {
+                return defaultValue;
+            }
+
+            switch (str.ToUpperInvariant())
+            {
+                case "TRUE":
+                case "1":
+                case "YES":
+                    return true;
+                case "FALSE":
+                case "0":
+                case "NO":
+                    return false;
+                default:
+                    return defaultValue;
+            }
+        }
+
+        private bool PermissionExists(IFirstTimeUseNoticeSentinel sentinel)
+        {
+            if (sentinel == null)
+            {
+                return false;
+            }
+
+            return sentinel.Exists();
+        }
+
+        public void TrackEvent(string eventName, IReadOnlyDictionary<string, string> properties,
+            IReadOnlyDictionary<string, double> measurements)
+        {
+            if (!Enabled)
+            {
+                return;
+            }
+
+            //continue task in existing parallel thread
+            _trackEventTask = _trackEventTask.ContinueWith(
+                x => TrackEventTask(eventName, properties, measurements),
+                TaskScheduler.Default
+            );
+        }
+
+        private void ThreadBlockingTrackEvent(string eventName, IReadOnlyDictionary<string, string> properties, IReadOnlyDictionary<string, double> measurements)
+        {
+            if (!Enabled)
+            {
+                return;
+            }
+            TrackEventTask(eventName, properties, measurements);
+        }
+
+        private void InitializeTelemetry(string productVersion)
+        {
+            try
+            {
+#pragma warning disable CS0618 // Type or member is obsolete
+                _client = new TelemetryClient();
+#pragma warning restore CS0618 // Type or member is obsolete
+                _client.InstrumentationKey = InstrumentationKey;
+                _client.Context.Session.Id = CurrentSessionId;
+                _client.Context.Device.OperatingSystem = RuntimeEnvironment.OperatingSystem;
+
+                _commonProperties = new TelemetryCommonProperties(productVersion).GetTelemetryCommonProperties();
+                _commonMeasurements = new Dictionary<string, double>();
+            }
+            catch (Exception e)
+            {
+                _client = null;
+                // we dont want to fail the tool if telemetry fails.
+                Debug.Fail(e.ToString());
```

Comment:
I added #388 because I think that is something that could be useful even outside of this narrow scenario. But we don't have any logging to disk right now at all, so I didn't want to try to cram that in here.